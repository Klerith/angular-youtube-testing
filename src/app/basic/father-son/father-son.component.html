<h5>Hijo</h5>
<code>{{ client | json }}</code>

<br><br>

<button
    *ngIf="client"
    class="btn btn-outline-primary"
    (click)="onChange(5)"
    data-test="btn-id"
>
    Cambiar ID
</button>


<button
    *ngIf="client"
    class="btn btn-outline-danger"
    (click)="onDelete()"
    data-test="btn-delete"
>
    Borrar cliente
</button>