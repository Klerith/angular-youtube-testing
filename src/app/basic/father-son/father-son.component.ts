import { Component, EventEmitter, Input, Output } from '@angular/core';
import { Client } from '../interfaces';

@Component({
  selector: 'app-father-son',
  templateUrl: './father-son.component.html',
  styleUrls: ['./father-son.component.css']
})
export class FatherSonComponent  {

  @Input() client?: Client;
  @Output() onDeleteClient  = new EventEmitter();
  @Output() onClientUpdated = new EventEmitter<Client>();

  onDelete() {
    this.client = undefined;
    this.onDeleteClient.emit();
  }

  onChange( id: number ) {
    if ( !this.client ) return;
    
    // this.client.id = newId; // no usar
    this.client = { ...this.client, id };

    this.onClientUpdated.emit({ ...this.client })

  }


}
