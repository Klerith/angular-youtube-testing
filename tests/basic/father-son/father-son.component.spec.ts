import { ComponentFixture, TestBed } from '@angular/core/testing';

import { FatherSonComponent } from '../../../src/app/basic/father-son/father-son.component';

describe('FatherSonComponent', () => {
  let component: FatherSonComponent;
  let fixture: ComponentFixture<FatherSonComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [ FatherSonComponent ]
    })
    .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(FatherSonComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
    jest.clearAllMocks();
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('debe de hacer match con el snapshot', () => {
    expect( compiled ).toMatchSnapshot();    
  });

  test('no deben de aparecer botones si no hay cliente', () => {
    const buttons = compiled.querySelectorAll('button');
    expect( buttons.length ).toBe(0);
  })

  test('deben de aparecer 2 botones si HAY cliente', () => { 
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    const buttons = compiled.querySelectorAll('button');
    expect( buttons.length ).toBe(2);
  });

  test('si hay cliente, hacer match con el snapshot', () => { 
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();
    expect( compiled ).toMatchSnapshot();
  });

  test('debe de emitir onDeleteClient con el boton de elmininar', () => {

    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    jest.spyOn( component.onDeleteClient, 'emit' );

    const btnDelete = compiled.querySelector('[data-test=btn-delete]');
    btnDelete?.dispatchEvent( new Event('click') );

    expect( component.onDeleteClient.emit ).toHaveBeenCalled();
    
  });

  test('debe de emitir onClientUpdated con el boton de "Cambiar ID"', () => {

    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    jest.spyOn( component.onClientUpdated, 'emit' );

    const btnChangeId = compiled.querySelector('[data-test=btn-id]');
    btnChangeId?.dispatchEvent( new Event('click') );

    expect( component.onClientUpdated.emit ).toHaveBeenCalledWith({
      id: 5,
      name: 'Juan'
    });
    
  });


  test('deben de emitir onChangeClient con el ID especificado SI hay un cliente', () => {

    jest.spyOn( component.onClientUpdated, 'emit' );
    component.onChange(10);
    expect( component.onClientUpdated.emit ).not.toHaveBeenCalled();

    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();
    component.onChange(10);

    expect( component.onClientUpdated.emit ).toHaveBeenCalledWith({
      id: 10,
      name: 'Juan'
    });

  });

    



});
