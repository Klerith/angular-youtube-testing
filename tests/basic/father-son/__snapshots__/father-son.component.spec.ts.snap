// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherSonComponent debe de hacer match con el snapshot 1`] = `
<div
  id="root1"
>
  <h5>
    Hijo
  </h5>
  <code>
    
  </code>
  <br />
  <br />
  
  
</div>
`;

exports[`FatherSonComponent si hay cliente, hacer match con el snapshot 1`] = `
<div
  id="root4"
>
  <h5>
    Hijo
  </h5>
  <code>
    {
  "id": 1,
  "name": "Juan"
}
  </code>
  <br />
  <br />
  <button
    class="btn btn-outline-primary"
    data-test="btn-id"
  >
     Cambiar ID

  </button>
  
  <button
    class="btn btn-outline-danger"
    data-test="btn-delete"
  >
     Borrar cliente

  </button>
  
</div>
`;
